gio/src/subclass/action.rs

```diff
@@ -0,0 +1,427 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::sync::OnceLock;
+
+use glib::{
+    ffi::{gboolean, GVariant, GVariantType},
+    prelude::*,
+    subclass::prelude::*,
+    translate::*,
+    GString, Variant, VariantType,
+};
+
+use crate::{ffi, Action};
+
+pub trait ActionImpl: ObjectImpl + ObjectSubclass<Type: IsA<Action>> {
+    fn name(&self) -> GString {
+        self.parent_name()
+    }
+
+    fn parameter_type(&self) -> Option<VariantType> {
+        self.parent_parameter_type()
+    }
+
+    fn state_type(&self) -> Option<VariantType> {
+        self.parent_state_type()
+    }
+
+    fn state_hint(&self) -> Option<Variant> {
+        self.parent_state_hint()
+    }
+
+    fn is_enabled(&self) -> bool {
+        self.parent_enabled()
+    }
+
+    fn state(&self) -> Option<Variant> {
+        self.parent_state()
+    }
+
+    fn change_state(&self, value: Variant) {
+        self.parent_change_state(value);
+    }
+
+    fn activate(&self, parameter: Option<Variant>) {
+        self.parent_activate(parameter);
+    }
+
+    fn delegate_get_property(
+        &self,
+        prop_id: usize,
+        _pspec: &glib::ParamSpec,
+    ) -> Option<glib::Value> {
+        let type_: glib::Type = self.obj().type_();
+        let property = ActionProperty::from_type(type_, prop_id)?;
```

Comment:
If you pass in `Self::type_()` here then you don't need to do all the parent type walking inside `from_type()` and always have directly the correct type

Reply:
That's what I've found in `GtkEditable`'s [code](https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gtk/gtkeditable.c?ref_type=heads#L1078-1081).

Reply:
@bilelmoussaoui Why does it do things so complicated, do you remember? :)